### PR TITLE
refactor: union `public`/`client.host`/`client.port`/`client.path`

### PR DESCRIPTION
**Pull request by alexander-akait**
- [ ] This is a **bugfix**
- [ ] This is a **feature**
- [x] This is a **code refactor**
- [x] This is a **test update**
- [ ] This is a **docs update**
- [x] This is a **metadata update**

### For Bugs and Features; did you add new tests?

Shortly - union `public`/`client.host`/`client.port`/`client.path` into `client.webSocketUrl`

fixes #3169

### Motivation / Use-Case

Better DX, simplify logic

### Breaking Changes

Yes, options was removed, also you need put `firewall: ['domain.com']` when your proxy dev server (**always**)

### Additional Info

Possible values:
- `webSocketUrl: 'ws://0.0.0.0:0/ws'` - full specified URL, `0.0.0.0` for using `self.location.hostname`, `:0` for using `self.location.port`
- `webSocketUrl: 'wss://192.168.0.1:8082/custom'` - wss, custom hostname, custom port and custom pathname

Summary: 
- When you use `0.0.0.0`, `self.location.hostname` will be used
- When you use `:0` or don't specify it, `self.location.port` will be used
- Default `pathname` is `/ws`, we will improve supporting in future, if you don't `pathname` `/ws` will be used by default

### CODE REVIEW

File: lib/utils/DevServerPlugin.js [hunk 1]
```diff
@@ -62,7 +39,9 @@ class DevServerPlugin {
     /** @type {string} */
     const clientEntry = `${require.resolve(
       '../../client/'
-    )}?${domain}${host}${path}${port}${logging}`;
+    )}?${domain}${logging}`;
+
+    console.log(domain);
```
Comment (snitin315) [inline lib/utils/DevServerPlugin.js:44]:
```suggestion
```

File: lib/utils/createDomain.js [hunk 2]
```diff
@@ -2,30 +2,24 @@
...
-
-  const port = server ? server.address().port : 0;
 
   // use explicitly defined public url
   // (prefix with protocol if not explicitly given)
-  if (options.public) {
-    return /^[a-zA-Z]+:\/\//.test(options.public)
-      ? `${options.public}`
-      : `${protocol}://${options.public}`;
+  if (options.client.socketUrl) {
```
Comment (snitin315) [inline lib/utils/createDomain.js:10]:
```suggestion
  if (options.client && options.client.socketUrl) {
```

Comment (alexander-akait) [reply lib/utils/createDomain.js:10]: We set `options.client` to `{}` in `normalizeOptions`

Comment (snitin315) [reply lib/utils/createDomain.js:10]: All good then 👍🏻

**Comment by snitin315** [comment]
Also, we need to remove `--public` in favor of `--client-socket-url`.

https://github.com/webpack/webpack-dev-server/blob/c4279c20f2cd6a02f60f7d1114e90808a46f8fda/bin/cli-flags.js#L234

**Comment by alexander-akait** [comment]
Yep, still WIP, we need more things here

**Comment by alexander-akait** [comment]
@snitin315 I think maybe we need to use `webSocketUrl` instead `socketUrl`, what do you think?

**Comment by snitin315** [comment]
Yes, let's go with `webSocketUrl` 👍.

**Comment by alexander-akait** [comment]
We need fix line https://github.com/webpack/webpack-cli/blob/master/packages/serve/src/startDevServer.ts#L77 in `webpack-cli`, we should not set port, because by default we use `0`, i.e. `self.location.port`, but we need to use this only for current version (this PR), we can check this on webpack-cli side like `const isModernClient = devServerCliOptions['client-web-socker-url']` (pseudo code) and do not set `client.port` if `isModernClient` is `true`

**Comment by snitin315** [comment]
> We need fix line https://github.com/webpack/webpack-cli/blob/master/packages/serve/src/startDevServer.ts#L77 in `webpack-cli`, we should not set port now, because by default we use `0`, i.e. `self.location.port`, but we need to use this only for current version (this PR), we can check this on webpack-cli side like `const isModernClient = devServerCliOptions['client-web-socker-url']` (pseudo code) and do not set `client.port` if `isModernClient` is `true`

I will take care of it

File: lib/Server.js [hunk 3]
```diff
@@ -937,19 +937,6 @@ class Server {
...
-      const idxPublic = this.options.public.indexOf(':');
-      const publicHostname =
-        idxPublic >= 0
-          ? this.options.public.substr(0, idxPublic)
-          : this.options.public;
-
-      if (hostname === publicHostname) {
-        return true;
-      }
-    }
```
Comment (alexander-akait) [inline lib/Server.js:951]: We have mixed logic for proxied server and socket url, it is weird, when you proxy dev server you need specify `host` in `firewall`

File: lib/servers/SockJSServer.js [hunk 4]
```diff
@@ -50,7 +50,7 @@ module.exports = class SockJSServer extends BaseServer {
     });
 
     this.socket.installHandlers(this.server.server, {
-      prefix: this.server.options.client.path,
+      prefix: '/ws',
```
Comment (alexander-akait) [inline lib/servers/SockJSServer.js:53]: Will be improved in future

File: lib/servers/WebsocketServer.js [hunk 5]
```diff
@@ -12,7 +12,7 @@ module.exports = class WebsocketServer extends BaseServer {
 
     this.wsServer = new ws.Server({
       noServer: true,
-      path: this.server.options.client.path,
+      path: '/ws',
```
Comment (alexander-akait) [inline lib/servers/WebsocketServer.js:15]: Will be improved in future

File: test/server/clientOptions-option.test.js [hunk 6]
```diff
@@ -43,33 +37,6 @@ describe('client option', () => {
...
-    });
-
-    it('sets the sock path correctly and strips leading and trailing /s', () => {
-      expect(server.options.client.path).toEqual(path);
-    });
-
-    it('responds with a 200 second', (done) => {
-      req.get(path).expect(200, done);
-    });
-  });
```
Comment (alexander-akait) [inline test/server/clientOptions-option.test.js:71]: unnecessary test, we already have tests in `e2e`

**Comment by alexander-akait** [comment]
Maybe:
```ts
{
  client: {
    websocketTransport: 
      'ws' | 
      'sockjs' | 
      { 
        client: { type: 'ws' | string, options: Record<any, any> }, 
        server: { type: 'ws' | string, options: Record<any, any> } 
      }
  }
}
```

Allow to set options for websocket server:
```js
{
  client: { 
    websocketTransport: { 
      client: {  
        // type: 'ws', can be omitted, because default value is `ws`
        options: { socketUrl: 'ws://dev.site/custom' }
      },
      server: {  
        // type: 'ws', can be omitted, because default value is `ws`
        options: { path: '/custom' }
      }  
    }
  }
}
```

But multiple `client` is some misleading here... need think more, maybe anybody have idea.

**Comment by alexander-akait** [comment]
Done here https://github.com/webpack/webpack-dev-server/pull/3309

**Pull request closed without merging**